FAQ: Updating some of the answers.

shamisp opened the pull request:
As it was pointed out, some of the answer are out of date. Updating the
answers to reflect the current state of the code.

Signed-off-by: Pavel Shamis (Pasha) <pasharesearch@gmail.com>

## What
FAQ update

## Why ?
As it was pointed out, some of the answer are out of date. Updating the
answers to reflect the current state of the code.

## How ?
Answers have discussed

swx-jenkins3 commented:
Can one of the admins verify this patch?

shamisp commented:
Ok to test

yosefe commented on docs/source/faq.md:
```diff
@@ -61,7 +61,7 @@ Instead, GASNET can leverage UCX framework for fast end efficient implementation
 
 #### What is the relation between UCX and OFA Verbs or Libfabrics?
-UCX, is a middleware communication layer that relies on vendors provided user level drivers including OFA Verbs or libfabrics (or any other drivers provided by another communities or vendors) to implement high-level protocols which can be used to close functionality gaps between various vendors drivers including various libfabrics providers: coordination across various drivers, multi-rail capabil ...
+UCX is a middleware communication framework that relies on device drivers including RDMA, CUDA, ROCM APIs, etc. RDMA and OS-bypass network devices typically implement device drivers using the RDMA-core Linux subsystem that is supported by the UCX framework. Support for other network abstractions can be added based on requests and contributions from the community.
```
"typically implement device drivers using the RDMA-core Linux..." - sentence seems weird

shamisp replied:
Sounds okay to me but maybe @tonycurtis help improve this.

tonycurtis commented on docs/source/faq.md:
```diff
@@ -38,7 +38,7 @@ The UCX framework is maintained and supported by hardware vendors in addition to
 
 * **High level API for a broad range HPC programming models.**  
-UCX provides a high level API implemented in software 'UCP' to fill in the gaps across interconnects. This allows to use a single set of APIs in a library  to implement multiple interconnects. This reduces the level of complexities when implementing libraries such as Open MPI or OpenSHMEM.  Because of this, UCX performance portable  because a single implementation (in Open MPI or OpenSHMEM) will  ...
+UCX provides a high level API implemented in software 'UCP' to fill in the gaps across interconnects. This allows to use a single set of APIs in a library  to implement multiple interconnects. This reduces the level of complexities when implementing libraries such as Open MPI or OpenSHMEM.  Because of this, UCX performance portable  because a single implementation (in Open MPI or OpenSHMEM) will  ...
```
UCX provides a high level API implemented in software (UCP) to fill in the gaps across interconnects. This allows the use of a single set of APIs in a library to implement multiple interconnects, and reduces the level of complexity when implementing libraries such as Open MPI or OpenSHMEM.  UCX is therefore performance portable because a single implementation (in Open MPI or OpenSHMEM) will work efficiently on multiple interconnects. (e.g. uGNI, Verbs).

tonycurtis commented on docs/source/faq.md:
```diff
@@ -38,7 +38,7 @@ The UCX framework is maintained and supported by hardware vendors in addition to
+UCX provides a high level API implemented in software 'UCP' to fill in the gaps across interconnects. This allows to use a single set of APIs in a library  to implement multiple interconnects. This reduces the level of complexities when implementing libraries such as Open MPI or OpenSHMEM.  Because of this, UCX performance portable  because a single implementation (in Open MPI or OpenSHMEM) will  ...
 
 * **Support for interaction between multiple transports (or providers) to deliver messages.**  
 For example, UCX has the logic (in UCP) to make 'GPUDirect', IB' and share memory work together efficiently to deliver the data where is needed without the user dealing with this. 
```
shared memory

tonycurtis commented on docs/source/faq.md:
```diff
@@ -61,7 +61,7 @@ Instead, GASNET can leverage UCX framework for fast end efficient implementation
 
 #### What is the relation between UCX and OFA Verbs or Libfabrics?
-UCX, is a middleware communication layer that relies on vendors provided user level drivers including OFA Verbs or libfabrics (or any other drivers provided by another communities or vendors) to implement high-level protocols which can be used to close functionality gaps between various vendors drivers including various libfabrics providers: coordination across various drivers, multi-rail capabil ...
+UCX is a middleware communication framework that relies on device drivers including RDMA, CUDA, ROCM APIs, etc. RDMA and OS-bypass network devices typically implement device drivers using the RDMA-core Linux subsystem that is supported by the UCX framework. Support for other network abstractions can be added based on requests and contributions from the community.
```
UCX is a middleware communication framework that relies on device drivers, e.g. RDMA, CUDA, ROCM.  RDMA and OS-bypass network devices typically implement device drivers using the RDMA-core Linux subsystem that is supported by UCX.  Support for other network abstractions can be added based on requests and contributions from the community.

shamisp commented:
Fixes #6647

shamisp commented:
@tonycurtis I addressed your comments.

tonycurtis commented on docs/source/faq.md:
```diff
@@ -10,7 +10,7 @@ and relatively easy way to construct widely used HPC protocols: MPI tag matching
 
 #### What is UCP, UCT, UCS?
-* **UCT** is a transport layer that abstracts the differences across various hardware architectures and provides a low-level API that enables the implementation of communication protocols. The primary goal of the layer is to provide direct and efficient access to hardware network resources with minimal software overhead. For this purpose UCT relies on low-level drivers provided by vendors such as ...
+* **UCT** is a transport layer that abstracts the differences across various hardware architectures and provides a low-level API that enables the implementation of communication protocols. The primary goal of the layer is to provide direct and efficient access to hardware network resources with minimal software overhead. For this purpose UCT relies on low-level drivers provided by vendors such as ...
```
There's some unneeded and mis-formatted hyphenation here

shamisp replied:
@tonycurtis good catch. please take a look at latest.

tonycurtis commented on docs/source/faq.md:
```diff
@@ -61,7 +61,7 @@ Instead, GASNET can leverage UCX framework for fast end efficient implementation
+UCX is a middleware communication framework that relies on device drivers, e.g. RDMA, CUDA, ROCM. RDMA and OS-bypass network devices typically implement device drivers using the RDMA-core Linux subsystem that is supported by UCX. Support for other network abstractions can be added based on requests and contributions from the community.
 
 #### Is UCX a user level driver?
 No. Typically,  Drivers  aim to expose fine-grain access to the network architecture specific features.
```
Typically, drivers...

fine-grained

architecture-specific

shamisp replied:
ah, this was not part of the change but I will fix.

tonycurtis commented on docs/source/faq.md:
```diff
@@ -38,7 +38,7 @@ The UCX framework is maintained and supported by hardware vendors in addition to
 The framework design, data structures, and components are design to provide highly optimized access to the network hardware. 
```
"designed" and the earlier use of "design" is discordant.  I'd just remove the first use of "design".

"highly optimized" sounds a bit over-the-top, but if you want to plug UCX's abilities, OK :)

tonycurtis commented on docs/source/faq.md:
```diff
@@ -38,7 +38,7 @@ The UCX framework is maintained and supported by hardware vendors in addition to
+UCX provides a high level API implemented in software (UCP) to fill in the gaps across interconnects. This allows the use of a single set of APIs in a library to implement multiple interconnects, and reduces the level of complexity when implementing libraries such as Open MPI or OpenSHMEM. UCX is therefore performance portable because a single implementation (in Open MPI or OpenSHMEM) will work e ...
 
 * **Support for interaction between multiple transports (or providers) to deliver messages.**  
 For example, UCX has the logic (in UCP) to make 'GPUDirect', IB' and share memory work together efficiently to deliver the data where is needed without the user dealing with this. 
```
where it is needed ...

shamisp replied:
not part of the commit

shamisp commented:
@tonycurtis I addressed your comments. please take a look. thanks

tonycurtis commented:
> On Apr 28, 2021, at 3:04 PM, Pavel Shamis (Pasha) ***@***.***> wrote:
> 
> 
> @tonycurtis <https://github.com/tonycurtis> I addressed your comments. please take a look. thanks
> 

Looks like a number of suggested edits haven’t been addressed, e.g. across -> between, active message -> active messages


But making progress!

Tony

shamisp commented:
> On Apr 28, 2021, at 3:04 PM, Pavel Shamis (Pasha) ***@***.***> wrote: @tonycurtis <https://github.com/tonycurtis> I addressed your comments. please take a look. thanks
> Looks like a number of suggested edits haven’t been addressed, e.g. across -> between, active message -> active messages But making progress! Tony

@tonycurtis you confuse with another PR on FAQ.

tonycurtis commented on docs/source/faq.md:
```diff
@@ -61,10 +61,10 @@ Instead, GASNET can leverage UCX framework for fast end efficient implementation
 
 #### Is UCX a user level driver?
-No. Typically,  Drivers  aim to expose fine-grain access to the network architecture specific features.
+UCX is not a user level driver. Typically, drivers  aim to expose fine-grained access to the network architecture-specific features.
```
user-level, and double space after "drivers"

shamisp replied:
fixed

shamisp commented:
@tonycurtis good to go ?

shamisp commented:
@yosefe seem like the CI miss-behaves

shamisp commented:
@tonycurtis  can you please do GitHub approve

shamisp commented:
@yosefe seems like you need approve as well.
